generic-ec-curves/src/utils.rs

```diff
@@ -0,0 +1,270 @@
+use generic_ec_core::Reduce;
+
+/// Interprets `bytes` as little-endian encoding of an integer, takes it modulo curve (prime)
+/// order and returns scalar `S`
+///
+/// Works with scalars for which both [`Reduce<32>`][Reduce] and [`Reduce<64>`][Reduce] are
+/// defined.
+///
+/// Takes:
+/// * Little-endian `bytes` representation of the integer
+/// * Scalar `one = 1`
+pub fn scalar_from_le_bytes_mod_order_reducing_32_64<S>(bytes: &[u8], one: &S) -> S
+where
+    S: Default + Copy,
+    S: Reduce<32> + Reduce<64>,
+    S: generic_ec_core::Additive + generic_ec_core::Multiplicative<S, Output = S>,
+{
+    let len = bytes.len();
+    match len {
+        ..=31 => {
+            let mut padded = [0u8; 32];
+            padded[..len].copy_from_slice(bytes);
+            S::from_le_array_mod_order(&padded)
+        }
+        32 => {
+            #[allow(clippy::expect_used)]
+            let bytes: &[u8; 32] = bytes.try_into().expect("we checked that bytes len == 32");
+            S::from_le_array_mod_order(bytes)
+        }
+        33..=63 => {
+            let mut padded = [0u8; 64];
+            padded[..len].copy_from_slice(bytes);
+            S::from_le_array_mod_order(&padded)
+        }
+        64 => {
+            #[allow(clippy::expect_used)]
+            let bytes: &[u8; 64] = bytes.try_into().expect("we checked that bytes len == 64");
+            S::from_le_array_mod_order(bytes)
+        }
+        65.. => {
+            let two_to_512 = S::add(&S::from_le_array_mod_order(&[0xff; 64]), one);
```

Comment:
Maybe
```
let mut bytes = [0; 65];
bytes[0] = 1;
S::from_le_array_mod_order(&bytes);
```

Reply:
Also, isn't this two_to_512?

Reply:
It will fall into infinite recursion :smiley:

Reply:
or I misunderstand the code. `S::from_le_array_mod_order(&bytes)` requires that `Reduce<65>` is implemented for the scalar. There's no scalar which implements it

Reply:
Ah right, my bad